### ISSUE TITLE
EN-DE and DE-EN backtranslations

### ISSUE THREAD

**Issue by nithya4** [pytorch/fairseq#524]
I am trying to generate alternates for sentences by translating the English source sentences into German and back. 

EN-DE:
I used the pretrained EN-DE transformer model provided in the documentation with n-best set to 5 (generating 5 variations). I use the generation scripts exactly as described [here.](https://github.com/pytorch/fairseq/blob/master/examples/translation/README.md)

DE-EN:
I trained a transformer model (transformer_iwslt_de_en) on IWSLT14 data using BPE(32000 tokens). This data is smaller than the WMT16 data used above. I get a BLEU of about 31 on the test set. 
I then use this model to translate the 5 best EN-DE hypotheses to EN with n-best 5. 

In this process, I have encountered some issues.
1) In the EN-DE Translation, using the pretrained model, I noticed an absurd occurrence:
```
S-6424	i think she &apos;s getting too thin , but if i talk to her about this , she get worried .
T-6424	
H-6424	-0.3724592626094818	I think she &apos;s getting too thin , but if i talk with her about this , she get worried .
H-6424	-0.39234718680381775	I think she &apos;s getting too thin , but if i talk with her about this , she get worried .
H-6424	-0.39331480860710144	I think she &apos;s getting too thin , but if i talk with her about this , she get concerned .
H-6424	-0.4266684353351593	I think she &apos;s getting too thin , but if i talk with her about this , she get concerned .
H-6424	-0.4446612596511841	I think she &apos;s getting too thin , but if i talk with her about this , she get worry .
```

The majority of responses seem to be in the target language but a few source sentences like these have responses in English. I've noticed a few more sentences with mixed EN+DE phrases. I am not sure why this is happening. 

2) The DE-EN model generalizes very poorly. I suspect that the size of the IWSLT dataset may be the issue. Any suggestions to improve the quality of responses? 

3) As an initial experimentation, I used both models independent of each other (different BPE codes). On sampling a few EN-DE responses, I found them to be of good quality in responses. 
However, an independent run of the DE-EN model (trained on IWSLT) does not produce desirable responses. I have also experimented with varying numbers of BPE tokens. 
Is it better off to retrain the DE-EN model on the WMT16 dataset, or apply the learned BPE from the WMT16 dataset on the IWSLT data? 

4) With regards to the response in #458, is using`--joined-dictionary` arg during preprocessing helpful for generation in both directions? On these datasets, what is the expected behavior for ` --share-all-embeddings` (when vocab is shared) and `--share-decoder-input-output-embed` options (when vocab is not shared) during training of the DE-EN model?

**Comment by semsomi** [comment]
Regarding 1, I'm also getting some untranslated sentences. It's just printing out the source sentences with minor variations. I've tried different args but can't seem to fix the issue.

**Comment by myleott** [comment]
I've added a more detailed README for reproducing the BT results from the Understanding BT at Scale paper: https://github.com/pytorch/fairseq/tree/master/examples/backtranslation#training-your-own-model-wmt18-english-german
